### PR TITLE
Update ImageExtension to work with Managed Disks

### PR DESCRIPTION
**Pull request by nacx**
This PR updates the ImageExtension to work with managed disks. The flow changes a bit since Managed VMs don't need to be captured to a blob before creating an image from them. In this PR I've also:

* Removed all uses of the BlobHelper and the dependency on azure blob.
* Fixed the annotations in several API classes.
* Updated the live tests. All should be passing now.

### CODE REVIEW

**Comment by andreaturli** [comment]
Thanks @nacx looks a very useful simplification: it's quite a big PR so I need to allocate some time to review carefully. I'll report the results of live tests asap, meanwhile

**Comment by andreaturli** [comment]
hi @nacx trying to run the live tests with maven but after 2 hours it is still at 
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
Running TestSuite
Configuring TestNG with: TestNG652Configurator
Starting test initializeContext(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
[TestNG] Test initializeContext(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) succeeded: 72ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
Starting test setServiceDefaults(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
[TestNG] Test setServiceDefaults(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) succeeded: 0ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
Starting test testCompareSizes(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
Apr 21, 2017 4:22:30 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest doCompareSizes
INFO: smallest {id=Standard_A0, providerId=Standard_A0, name=Standard_A0, location={scope=REGION, id=westeurope, description=West Europe, parent=azurecompute-arm, iso3166Codes=[NL]}, processors=[{cores=1.0, speed=2.0}], ram=768, volumes=[{type=LOCAL, size=20480.0, bootDevice=false, durable=false}, {type=LOCAL, size=1047552.0, bootDevice=false, durable=false}], supportsImage=ALWAYS_TRUE, userMetadata={maxDataDiskCount=1}}
Apr 21, 2017 4:22:30 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest doCompareSizes
INFO: fastest {id=Standard_GS5, providerId=Standard_GS5, name=Standard_GS5, location={scope=REGION, id=westeurope, description=West Europe, parent=azurecompute-arm, iso3166Codes=[NL]}, processors=[{cores=32.0, speed=2.0}], ram=458752, volumes=[{type=LOCAL, size=917504.0, bootDevice=false, durable=false}, {type=LOCAL, size=1047552.0, bootDevice=false, durable=false}], supportsImage=ALWAYS_TRUE, userMetadata={maxDataDiskCount=64}}
Apr 21, 2017 4:22:30 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest doCompareSizes
INFO: biggest {id=Standard_G5, providerId=Standard_G5, name=Standard_G5, location={scope=REGION, id=westeurope, description=West Europe, parent=azurecompute-arm, iso3166Codes=[NL]}, processors=[{cores=32.0, speed=2.0}], ram=458752, volumes=[{type=LOCAL, size=6291456.0, bootDevice=false, durable=false}, {type=LOCAL, size=1047552.0, bootDevice=false, durable=false}], supportsImage=ALWAYS_TRUE, userMetadata={maxDataDiskCount=64}}
[TestNG] Test testCompareSizes(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) succeeded: 27161ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testCorrectAuthException(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
[TestNG] Test testCorrectAuthException(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) succeeded: 600ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testCorrectExceptionRunningNodesNotFound(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
[TestNG] Test testCorrectExceptionRunningNodesNotFound(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) succeeded: 85ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Starting test testCreateAndRunAService(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
Apr 21, 2017 4:26:33 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest createAndRunAServiceInGroup
INFO: << available node(westeurope/azurecompute-arms-e54) os({family=ubuntu, version=14.04.5-LTS, description=14.04.5-LTS, is64Bit=true}) in 242s
Apr 21, 2017 4:27:08 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest createAndRunAServiceInGroup
INFO: << configured node(westeurope/azurecompute-arms-e54) with bash: java: command not found and jetty head: cannot open ‘/usr/local/jetty/VERSION.txt’ for reading: No such file or directory in 12s
```
is it expected?

**Comment by nacx** [comment]
I'm working on fixing this. I still don't understand why it hangs. If you run the `testCreateAndRunAService` alone it succeeds, but when running the entire class it hangs there.

I've looking at this for two days and haven't been able to figure it out yet. The code hangs after starting Jetty. The test tries to connect to the Jetty port using the jclouds ConcurrentOpenSocketFinder, but for some reason, I've observed that in the tests, in that concrete case, the OpenSocketFinder implementation has an instance of the "user executor" that is already terminated, thus the tasks to check for the port remain enqueued forever. Here is what I'm investigating now:

1. Why there is a different instance of the user executor in there, when it is declared as a Singleton in the ExecutorService.
2. I'm trying to debug when the executor service is shut down, but still haven't been able to determine it.
3. I'm studying the need to use a custom executor like [the default jclouds uses](https://github.com/jclouds/jclouds/blob/master/core/src/main/java/org/jclouds/concurrent/DynamicExecutors.java). I don't really know why using such an executor is needed, but the current implementation, re-enqueues the tasks when they are rejected, instead of throwing an exception. Removing that behaviour may unlock this tests, since an exception would be thrown when the socket check tasks are enqueued in an already terminated executor.

None of these changes seems to be related to the changes in this particular provider, but this is what I've found and the reason the code hangs forever. I'll keep you posted on my progress here!

**Comment by nacx** [comment]
Ok @andreaturli I think I've found the root cause of the issue: a misconfiguration of the base test class. Details here: https://github.com/jclouds/jclouds/pull/1088.
I'd pretty much appreciate if you could build that branch with the fix and run the tests again!

**Comment by nacx** [comment]
And I think I've addressed the root cause of the test hanging here: https://github.com/jclouds/jclouds/pull/1089

**Comment by andreaturli** [comment]
@nacx using your jclouds/jclouds#1089 seems to pass the jetty issue but the test itself fails
```
Starting test testCreateAndRunAService(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest)
Apr 21, 2017 10:58:54 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest createAndRunAServiceInGroup
INFO: << available node(westeurope/azurecompute-arms-ac4) os({family=ubuntu, version=16.04.0-LTS, description=16.04.0-LTS, is64Bit=true}) in 216s
Apr 21, 2017 11:02:46 PM org.jclouds.compute.internal.BaseComputeServiceLiveTest createAndRunAServiceInGroup
INFO: << configured node(westeurope/azurecompute-arms-ac4) with openjdk full version "9-internal+0-2016-04-14-195246.buildd.src" and jetty jetty-8.1.8.v20121106 in 221s
[TestNG] Test testCreateAndRunAService(org.jclouds.azurecompute.arm.compute.AzureComputeServiceLiveTest) failed.
Test suite progress: tests succeeded: 5, failed: 1, skipped: 0.
```

**Comment by nacx** [comment]
Running the entire test suite with all my PRs works fine:

```bash
Results :

Failed tests:
  AzureComputeSecurityGroupExtensionLiveTest.testCreateNodeWithInboundPorts:125 » IllegalArgument
  AzureComputeSecurityGroupExtensionLiveTest.testCreateNodeWithSecurityGroup:107 » IllegalArgument

Tests run: 143, Failures: 2, Errors: 0, Skipped: 2
```

There are 2 test failures due to a naming change in the test resource group names I'll fix in an upcoming commit, and the 2 skips are normal due to unsupported operations in the SecurityGroupExtension.

**Comment by nacx** [comment]
With the last commit in this PR and the latest 2.0.x all tests should be green!

File: azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/internal/AzureLiveTestUtils.java [hunk 1]
```diff
@@ -32,21 +32,20 @@
...
-       properties.setProperty(TIMEOUT_SCRIPT_COMPLETE, defaultTimeout);
-       properties.setProperty(TIMEOUT_NODE_RUNNING, defaultTimeout);
-       properties.setProperty(TIMEOUT_PORT_OPEN, defaultTimeout);
-       properties.setProperty(TIMEOUT_NODE_TERMINATED, defaultTimeout);
-       properties.setProperty(TIMEOUT_NODE_SUSPENDED, defaultTimeout);
-       
-       return properties;
-    }
-}
+   public static Properties defaultProperties(Properties properties, Object currentTestClass) {
```
Comment (andreaturli) [inline azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/internal/AzureLiveTestUtils.java]: wouldn't be more explicit to modify `Object currentTestClass`to `String resourceNamePrefix` directly

**Comment by andreaturli** [comment]
my last run shows
```
Results :

Failed tests:
  AzureComputeSecurityGroupExtensionLiveTest.testCreateNodeWithInboundPorts:134 expected [{ipProtocol=tcp, fromPort=22, toPort=24, tenantIdGroupNamePairs={}, groupIds=[], cidrBlocks=[0.0.0.0/0], exclusionCidrBlocks=[]}] but found [{ipProtocol=tcp, fromPort=8000, toPort=8000, tenantIdGroupNamePairs={}, groupIds=[], cidrBlocks=[0.0.0.0/0], exclusionCidrBlocks=[]}]

Tests run: 143, Failures: 1, Errors: 0, Skipped: 2
```

much better, although *Total time: 03:24 h*!!!

**Comment by nacx** [comment]
It looks like a transient test failure. Tests took half the time for me, though. Creating some VMs takes quite a long time, but not that much. Have you checked that they ran only *once*? Depending on the maven goals you configure they run in the integration phase and in the verification phase.

Regarding the resource groups:

* The network is not removed if there are things attached there. Looks like the ClenaupResources strategy could do better to wait until everything is deleted to try deleting the network. However, if we want to properly support the `templateOptions.networkIds` (I'll start working on it in short), we wouldn't like to delete user's networks when undeploying the VMs, so I'd say it's not a big deal to leave the network there.
* Regarding the disks, you're right. I also observed that. The Managed Disks PR did not take into account the deletion of the managed disks in the CleanupResources strategy. It is a pending task not related to this PR that should be fixed apart.

**Comment by andreaturli** [comment]
ok @nacx makes a lot of sense. I've a minor comment still pending, if you can address it, the I think we can merge it

File: azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/internal/AzureLiveTestUtils.java (hunk 1 above)
Comment (nacx) [inline azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/internal/AzureLiveTestUtils.java]: Done!

**Comment by nacx** [comment]
I've just addressed the comment. Thanks for having a look and testing this!

**Comment by nacx** [comment]
Squashed and pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/9caaa469) and [2.0.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/ea62a4c3). Thanks @andreaturli!

**Pull request closed without merging**